mypy/checker.py

```diff
@@ -2383,6 +2384,35 @@ class C(B, A[int]): ...  # this is unsafe because...
         if not ok:
             self.msg.base_class_definitions_incompatible(name, base1, base2, ctx)
 
+    def check_metaclass_compatibility(self, typ: TypeInfo) -> None:
+        """Ensures that metaclasses of all parent types are compatible."""
+        if (
+            typ.is_metaclass()
+            or typ.is_protocol
+            or typ.is_named_tuple
+            or typ.is_enum
+            or typ.typeddict_type is not None
+        ):
+            return  # Reasonable exceptions from this check
+
+        metaclasses = [
+            entry.metaclass_type
+            for entry in typ.mro[1:-1]
+            if entry.metaclass_type
+            and not is_named_instance(entry.metaclass_type, "builtins.type")
+        ]
+        if not metaclasses:
+            return
+        if typ.metaclass_type is not None and all(
+            is_subtype(typ.metaclass_type, meta) for meta in metaclasses
+        ):
+            return
+        self.fail(
+            "Metaclass conflict: the metaclass of a derived class must be "
+            "a (non-strict) subclass of the metaclasses of all its bases",
+            typ,
+        )
```

Comment:
Maybe you could do something like this, to improve the error message:

```suggestion
        subclass_metaclass = typ.metaclass_type
        if subclass_metaclass is None:
            return
        for superclass in typ.mro[1:-1]:
            superclass_metaclass = superclass.metaclass_type
            if superclass_metaclass is None or is_named_instance(
                superclass_metaclass, "builtins.type"
            ):
                continue
            if not is_subtype(subclass_metaclass, superclass_metaclass):
                self.fail(
                    (
                        f"Metaclass conflict: metaclass {subclass_metaclass.fullname!r} "
                        f"is not a subclass of {superclass_metaclass.fullname!r}, "
                        f"the metaclass of base class {superclass.fullname!r}"
                    ),
                    typ
                )
```

Reply:
I've tried it and I didn't like it. Why?
Because it either reports several errors for several bases with incompatible metaclasses or reports just one (if `break` is added) which is potentially confusing or we need even more complex code to accumulate all problematic base types and then raise a single one.

Let's keep it as-is for now? 

If someone complains, then we can easily make this more readable and more complex.

(This is why I love coping CPython's error messages 😆 )

Reply:
> Let's keep it as-is for now?

Sure, the PR looks pretty good right now!